community/cypher/runtime-util/src/main/scala/org/neo4j/cypher/internal/runtime/LongArraySet.scala

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.runtime
+
+import java.lang.Integer.highestOneBit
+import java.util
+
+import org.neo4j.cypher.internal.runtime.LongArraySet._
+
+/**
+  * When you need to have a set of arrays of longs representing entities - look no further
+  *
+  * This set will keep all it's state in a single long[] array, marking unused slots
+  * using 0xF000000000000000L, a value that should never be used for node or relationship id's.
+  *
+  * The set will be resized when either probing has to go on for too long when doing inserts,
+  * or the load factor reaches 75%.
+  *
+  * @param capacity The initial capacity for the set. Must be a power of 2
+  * @param longsPerEntry All arrays in the set must be of this length
+  */
+class LongArraySet(capacity: Int = 32, longsPerEntry: Int) {
```

Comment:
It would be good to have better understanding of the performance characteristics of this implementation over a balanced range of data sets...